### PR TITLE
Add and test JSON object matcher

### PR DESCRIPTION
**Pull request by webmaster128**
JSON matching for #817

### CODE REVIEW

File: x/wasm/types/json_matching.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+import (
+	"encoding/json"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// isJSONObjectWithTopLevelKey checks if the given bytes are a valid JSON object
+// with exactly one top level key that is contained in the list of allowed keys.
+func isJSONObjectWithTopLevelKey(jsonBytes []byte, allowedKeys []string) error {
```
Comment (ethanfrey) [inline x/wasm/types/json_matching.go:11]: very nice function, doing what it should

File: x/wasm/types/json_matching.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	if len(document) > 1 {
+		return sdkerrors.Wrap(ErrMultipleTopLevelKeys, "JSON object has multiple top level keys")
+	}
+
+	// Loop is executed exactly once
+	for topLevelKey := range document {
+		found := false
+		for _, allowedKey := range allowedKeys {
+			if allowedKey == topLevelKey {
+				found = true
```
Comment (ethanfrey) [inline x/wasm/types/json_matching.go:30]: You could just `return nil` here

File: x/wasm/types/json_matching.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+			}
+		}
+
+		if !found {
+			return sdkerrors.Wrapf(ErrTopKevelKeyNotAllowed, "JSON object has a top level key which is not allowed: '%s'", topLevelKey)
```
Comment (ethanfrey) [inline x/wasm/types/json_matching.go]: And return this error here. Removing the `found` variable completely

File: x/wasm/types/errors.go [hunk 4]
```diff
@@ -72,6 +72,18 @@ var (
 
 	//  error if an address does not belong to a contract (just for registration)
 	_ = sdkErrors.Register(DefaultCodespace, 22, "no such contract")
+
+	// ErrNotAJSONObject error if given data is not a JSON object
+	ErrNotAJSONObject = sdkErrors.Register(DefaultCodespace, 23, "not a JSON object")
```
Comment (ethanfrey) [inline x/wasm/types/errors.go:77]: Good stuff with the unique error codes

File: x/wasm/types/json_matching_test.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+		},
+		"happy with whitespace": {
+			src: []byte(`{
+				"msg":	null    }`),
+			allowedKeys: []string{"msg"},
+			exp:         nil,
+		},
+		"happy with excaped key": {
+			src:         []byte(`{"event\u2468thing": {"foo":"bar"}}`),
+			allowedKeys: []string{"event⑨thing"},
```
Comment (ethanfrey) [inline x/wasm/types/json_matching_test.go:59]: 😄

File: x/wasm/types/json_matching_test.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
+package types
+
+import (
+	"testing"
+
+	// sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/stretchr/testify/require"
+)
+
+func TestIsJSONObjectWithTopLevelKey(t *testing.T) {
```
Comment (ethanfrey) [inline x/wasm/types/json_matching_test.go:10]: Wonderful test vectors here!

**Review by ethanfrey — COMMENTED** [review]
@jackzampolin you really should cherry-pick in these commits to your branch.
This is great example of testing methods

(Or maybe we just merge these two commits to main?)

File: x/wasm/types/json_matching.go (hunk 3 above)
Comment (webmaster128) [reply x/wasm/types/json_matching.go]: Right, done, thanks

**Comment by webmaster128** [comment]
Rebased onto main as this is independent of the diff in #817. Feel free to merge into main once @jackzampolin and maybe @alpe had a look.

**Review by ethanfrey — APPROVED** [review]
Happy to merge as soon as CI is happy (commented how to fix above)

**Pull request merged**